Add CodegenServerCluster: a ServerClusterInterface implementation for Ember/Codegen based clusters

soares-sergio opened the pull request:
This PR implements the `ServerClusterInterface` for the Codegen/Ember server clusters and updates the `CodegenDataModelProvider` to use this implementation.

This allows us to hide codegen/ember clusters under a generic interface, which will help us to develop and validate a future implementation of a `CodeDrivenDataModelProvider` that operates only on `ServerClusterInterface` clusters and doesn't depend on ember.

Some things to discuss:
* Need to decide if we want to allow `ServerClusterInterface` to have multiple `paths` and update the interface accordingly (for example, what should `GetPath()` return?). I already update `GetVersion()` to add a `path` argument.


#### Testing
Tested using unit tests.

* `TestCodegenModelViaMocks`:
  * `ninja -C out src/data-model-providers/codegen/tests:TestCodegenModelViaMocks`
  * ./out/tests/TestCodegenModelViaMocks
* `TestCodegenServerCluster`:
  * `ninja -C out/linux-x64-sergio/ src/data-model-providers/codegen/tests:TestCodegenServerCluster`
  * `/out/tests/TestCodegenServerCluster`

soares-sergio commented on src/data-model-providers/codegen/tests/EmberReadWriteOverride.cpp:
```diff
@@ -122,6 +123,13 @@ Status emAfWriteAttributeExternal(const chip::app::ConcreteAttributePath & path,
     gEmberIoBufferFill = len;
 
+    // Increase cluster data version
+    auto * version = emberAfDataVersionStorage(chip::app::ConcreteClusterPath(path.mEndpointId, path.mClusterId));
```
Noticed the mocks didn't increase the dataVersion and added it here. Could you double check if this makes sense @andy31415 ? Thanks!

andy31415 replied:
If we start testing that version increases, it is probably ok.

Ember has extra  logic to for "did binary content change" however we probably want to mock this only as far as we need tests to pass.

soares-sergio commented on src/app/server-cluster/ServerClusterInterface.h:
```diff
@@ -71,7 +71,7 @@ class ServerClusterInterface
     ///   [...]
     ///   A cluster data version SHALL be incremented if any attribute data changes.
-    [[nodiscard]] virtual DataVersion GetDataVersion() const = 0;
+    [[nodiscard]] virtual DataVersion GetDataVersion(const ConcreteClusterPath & path) const = 0;
```
Need to discuss if a single `ServerClusterInterface` can represent multiple paths, like what I'm doing here. If so, we need to update the Interface, for example, to have a path as argument here.

The interface is currently not consistent, some functions require a path as arguments while other don't.

soares-sergio replied:
Also, need to decide what to do with `ConcreteClusterPath GetPath()`. If it supports multiple paths, what are we returning here?

andy31415 replied:
#37931 is supposed to be mergeable (based on offline discussions, but still waiting for review from @bzbarsky-apple ) and that would allow multi-path. We could build the draft PR on the assumption that the PR merges.

bzbarsky-apple commented on src/data-model-providers/codegen/CodegenServerCluster.cpp:
So I have to ask the question again: is the goal here to get rid of the codegen provider, or are we keeping it?

Because if we are keeping it, there is no need for the stuff we're doing here, as far as I can tell: all that metadata can just keep living in generated bits that the codegen provider interacts with directly, no?

I'd really like to see a clear explanation of what the goal is here and how these changes are getting us closer to it.

soares-sergio committed:
Implement CodegenServerCluster based on the ServerClusterInterface

This PR implements the ServerClusterInterface for the Codegen/Ember
server clusters.

soares-sergio committed:
fix darwin xcode stuff

soares-sergio committed:
fix includes and im call invokeCommand from codegenServerCluster

soares-sergio committed:
fix missing return

soares-sergio commented:
Dude to the increase in flash size, I'll close this pull request. I'll use this on the side for the CodeDrivenProvider development.
